crates/typst/src/model/par.rs

```diff
@@ -206,3 +215,166 @@ impl ParbreakElem {
+    /// Typst is awesome.
+    /// ```
+    #[ghost]
+    pub numbering: Option<Numbering>,
+
+    /// The alignment of line numbers associated with each line.
+    ///
+    /// The default of `auto` will provide a smart default where numbers grow
+    /// horizontally away from the text, considering the margin they're in and
+    /// the current text direction.
+    ///
+    /// ```example
+    /// #set par.line(numbering: "I", number-align: left)
+    ///
+    /// Hello world! \
+    /// Today is a beautiful day \
+    /// For exploring the world.
+    /// ```
+    #[ghost]
+    pub number_align: Smart<HAlignment>,
+
+    /// The margin at which the line numbers for each line appear.
+    ///
+    /// ```example
+    /// #set par.line(numbering: "1", number-margin: right)
+    ///
+    /// = Report
+    /// - Brightness: Dark, yet darker
+    /// - Readings: Negative
+    /// ```
+    #[ghost]
+    #[default(OuterHAlignment::Start)]
+    pub number_margin: OuterHAlignment,
+
+    /// The distance between line numbers and text.
+    ///
+    /// ```example
+    /// #set par.line(
+    ///   numbering: "1",
+    ///   number-clearance: 0.5pt
+    /// )
+    ///
+    /// Typesetting \
+    /// Styling \
+    /// Layout
+    /// ```
+    #[ghost]
+    #[default(Length::from(Abs::cm(1.0)))]
+    pub number_clearance: Length,
+
+    /// Controls when to reset line numbering.
+    ///
+    /// Possible options are `"document"`, indicating the line number counter
+    /// is never reset, or `"page"`, indicating it is reset on every page.
+    ///
+    /// ```example
+    /// #set par.line(
+    ///   numbering: "1.",
+    ///   numbering-scope: "page",
+    /// )
+    ///
+    /// First line \
+    /// Second line
+    /// #pagebreak()
+    /// First line again
+    /// Second line again
+    /// ```
+    #[ghost]
+    #[default(ParLineNumberingScope::Document)]
+    pub numbering_scope: ParLineNumberingScope,
+}
+
+impl Construct for ParLine {
+    fn construct(_: &mut Engine, args: &mut Args) -> SourceResult<Content> {
+        bail!(args.span, "cannot be constructed manually");
+    }
+}
+
+impl Count for Packed<ParLine> {
+    fn update(&self) -> Option<CounterUpdate> {
+        // The line counter must be updated manually by the root flow
+        None
+    }
+}
+
+/// Possible line numbering scope options, indicating how often the line number
+/// counter should be reset.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum ParLineNumberingScope {
+    /// Indicates the line number counter spans the whole document, that is,
+    /// is never automatically reset.
+    Document,
+    /// Indicates the line number counter should be reset at the start of every
+    /// new page.
+    Page,
+}
+
+cast! {
+    ParLineNumberingScope,
+    self => match self {
+        Self::Document => "document".into_value(),
+        Self::Page => "page".into_value(),
+    },
+    "document" => Self::Document,
+    "page" => Self::Page,
+}
```

Comment:
You can use the `Cast` derive macro for this. Then the doc comments will also show up the in rendered Typst docs.